InterruptedException in CI run

InterruptedException in CI runs, macos-latest, 17:

* https://github.com/Locke/coreasm.core/actions/runs/3165778362/jobs/5155060006
* https://github.com/Locke/coreasm.core/actions/runs/3165777925/jobs/5155058423

```
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.53 s <<< FAILURE! - in org.coreasm.engine.test.plugins.schedulingpolicies.TestSchedul1
Error:  org.coreasm.engine.test.plugins.schedulingpolicies.TestSchedul1.performTest  Time elapsed: 2.139 s  <<< FAILURE!
java.lang.AssertionError: Test failed for class: TestAllCasm
```

```
An error occurred after 6 steps in TestSchedul1.casm: An error occurred in TestSchedul1.casm:java.lang.InterruptedException
	at java.base/java.util.concurrent.locks.ReentrantLock$Sync.lockInterruptibly(ReentrantLock.java:159)
	at java.base/java.util.concurrent.locks.ReentrantLock.lockInterruptibly(ReentrantLock.java:372)
	at java.base/java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:432)
	at org.coreasm.engine.Engine$EngineThread.processNextCommand(Engine.java:1141)
	at org.junit.Assert.fail(Assert.java:89)
	at org.coreasm.engine.Engine$EngineThread.run(Engine.java:820)
	at org.coreasm.engine.test.TestAllCasm.performTest(TestAllCasm.java:180)

	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:377)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:284)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:248)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:167)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
```

--- a/org.coreasm.engine/src/org/coreasm/engine/Engine.java
+++ b/org.coreasm.engine/src/org/coreasm/engine/Engine.java
@@ -1119,8 +1119,8 @@ public class Engine implements ControlAPI {
 		 *
 		 * @see #next(EngineMode)
 		 */
-		private void processNextCommand() throws EngineException, InterruptedException {
-			EngineCommand cmd;
+		private void processNextCommand() throws EngineException {
+			EngineCommand cmd = null;
 			int rrc = remainingRunCount.getAndDecrement();
 
 			if (rrc > 0)
@@ -1139,12 +1139,18 @@ public class Engine implements ControlAPI {
 				finally {
 					isBusyLock.unlock();
 				}
-
-				// blocking wait for the next command
-				cmd = commandQueue.take();
 			}
 
-			assert engineBusy;
+			while (cmd == null) {
+				try {
+					// blocking wait for the next command
+					cmd = commandQueue.take();
+					assert engineBusy;
+				}
+				catch (InterruptedException ex) {
+					logger.warn("Engine got interrupted while waiting for the next command. Going to wait again.");
+				}
+			}
 
 			lastCommand = cmd;
 

--- a/org.coreasm.engine/src/org/coreasm/engine/Engine.java
+++ b/org.coreasm.engine/src/org/coreasm/engine/Engine.java
@@ -996,10 +996,18 @@ public class Engine implements ControlAPI {
 									isBusyLock.unlock();
 								}
 
-								EngineCommand cmd = commandQueue.take();
-								assert engineBusy;
+								EngineCommand cmd = null;
+								try {
+									cmd = commandQueue.take();
+									assert engineBusy;
+								}
+								catch (InterruptedException ex) {
+									logger.warn("Engine is in error state and needs to be terminated or reset. Interrupted while waiting for TERMINATE or RECOVER command. Going to wait again.");
+								}
 
-								if (cmd.type == EngineCommand.CmdType.ecTerminate) {
+								if (cmd == null) {
+									// can only happen via InterruptedException, repeat the loop
+								} else if (cmd.type == EngineCommand.CmdType.ecTerminate) {
 									next(EngineMode.emTerminating);
 									lastError = null;
 									logger.debug("Engine terminated by user command.");

--- a/org.coreasm.engine/src/org/coreasm/engine/Engine.java
+++ b/org.coreasm.engine/src/org/coreasm/engine/Engine.java
@@ -793,15 +793,17 @@ public class Engine implements ControlAPI {
 		public void run() {
 			try {
 				while (!terminating) {
-					assert engineBusy;
+					assert engineBusy || (lastError != null);
 
 					try {
 						EngineMode engineMode = getEngineMode();
 
 						// if an error is occurred and the engine is not
 						// in error mode, go to the error mode
-						if (lastError != null
-								&& engineMode != EngineMode.emError) {
+						if (lastError != null && engineMode != EngineMode.emError) {
+							// NOTE: isBusyLock is only needed when switching to false (in order to signal isNotBusy),
+							// but not when switching to true, like here. Locking here might even be harmful.
+							engineBusy = true;
 							next(EngineMode.emError);
 						}
 

--- a/org.coreasm.engine/src/org/coreasm/engine/Engine.java
+++ b/org.coreasm.engine/src/org/coreasm/engine/Engine.java
@@ -1021,17 +1021,15 @@ public class Engine implements ControlAPI {
 
 						}
 					} catch (CoreASMError ce) {
-						error(ce);
 						logger.error("Error occurred: {}", ce.showError());
+						error(ce);
 					} catch (Exception e) {
+						logger.error("Exception occurred. ", e);
+
 						if (e instanceof ParserException)
 							error(new CoreASMError((ParserException)e));
 						else
 							error(e);
-						logger.error("Exception occurred. ", e);
-						// StackTraceElement[] trace = e.getStackTrace();
-						// for (StackTraceElement ste: trace)
-						//   logger.error( ste.toString());
 					}
 				}
 

--- a/org.coreasm.engine/test-rsc/engine_states/01-interrupt.casm
+++ b/org.coreasm.engine/test-rsc/engine_states/01-interrupt.casm
@@ -1,0 +1,44 @@
+// initialize
+// @check engineStatus emIdle
+// @do waitWhileBusy
+// @check engineStatus emIdle
+
+// softInterrupt: should only clear queue, not send an actual interrupt. Nothing should happen.
+// @do softInterrupt
+// @do sleep1s
+// @check engineStatus emIdle
+
+// hardInterrupt: send first interrupt. This should interrupt `commandQueue.take()` in `Engine.processNextCommand()`.
+// @do hardInterrupt
+// @do sleep1s
+
+// hardInterrupt: send second interrupt.
+// If first interruption was unhandled, then this should interrupt `commandQueue.take()` in `case emError`.
+// Otherwise, it will just do the same as the first interrupt.
+// @do hardInterrupt
+// @do sleep1s
+
+// hardInterrupt: send third interrupt. Just to be sure.
+// @do hardInterrupt
+// @do sleep1s
+// @check engineStatus emIdle
+
+// NOTE: no step
+
+// @refuse "STEP 0: lastStep = undef"
+
+CoreASM SingleStep
+
+use StandardPlugins
+use Time
+use Number
+use String
+
+function lastStep : -> NUMBER
+
+init initRule
+
+rule initRule = {
+	lastStep := stepcount
+	print "STEP " + stepcount + ": lastStep = " + lastStep
+}

--- a/org.coreasm.engine/test/org/coreasm/engine/test/DetailedTestCase.java
+++ b/org.coreasm.engine/test/org/coreasm/engine/test/DetailedTestCase.java
@@ -9,7 +9,7 @@ import org.coreasm.engine.CoreASMEngine;
 public class DetailedTestCase extends TestCase {
 
 	public enum TestCaseStepDoType {
-		waitWhileBusy, enqueueStep
+		waitWhileBusy, enqueueStep, softInterrupt, hardInterrupt, sleep1s
 	}
 
 	public enum TestCaseStepCheckType {

--- a/org.coreasm.engine/test/org/coreasm/engine/test/TestEngineStates.java
+++ b/org.coreasm.engine/test/org/coreasm/engine/test/TestEngineStates.java
@@ -120,6 +120,15 @@ public class TestEngineStates {
 							steps++;
 							td.engine.enqueueStep();
 							break;
+						case softInterrupt:
+							td.engine.softInterrupt();
+							break;
+						case hardInterrupt:
+							td.engine.hardInterrupt();
+							break;
+						case sleep1s:
+							Thread.sleep(1000);
+							break;
 					}
 				}
 				else if (testCaseStep instanceof DetailedTestCase.TestCaseStepCheck) {

